internal/functions/serve/watcher.go

```diff
@@ -0,0 +1,316 @@
+			absFunctionsPath = filepath.Join(utils.CurrentDirAbs, functionsDir)
+		} else {
+			// Fallback to current working directory
+			cwd, CWDerr := os.Getwd()
+			if CWDerr != nil {
+				utils.Warning("could not get current working directory: %v. Using relative path for watcher: %s", CWDerr, functionsDir)
+			} else {
+				absFunctionsPath = filepath.Join(cwd, functionsDir)
+			}
+		}
+	}
+	absFunctionsPath = filepath.Clean(absFunctionsPath)
+
+	// Add all import dependencies to the watcher (including the functions directory itself)
+	if err := addImportDependenciesToWatcher(watcher, watchedDirs, fsys); err != nil {
+		utils.Warning("could not add import dependencies to watcher: %v", err)
+		// Return the watcher but an empty path to signal that watching isn't properly set up.
+		return watcher, "", nil
+	}
+
+	// If no directories are being watched (e.g., functions directory doesn't exist), return empty path
+	if len(watchedDirs) == 0 {
+		utils.Info(1, "No directories found to watch")
+		return watcher, "", nil
+	}
+
+	utils.Info(1, "File watcher initialized, watching %d directories", len(watchedDirs))
+	return watcher, absFunctionsPath, nil
+}
+
+// runFileWatcher listens for events from the watcher, debounces them, and signals for a restart.
+func runFileWatcher(ctx context.Context, watcher *fsnotify.Watcher, watchedPath string, restartChan chan<- struct{}, watchedDirs map[string]bool, fsys afero.Fs) {
+	var restartTimer *time.Timer // Timer for debouncing restarts
+
+	if watchedPath == "" {
+		return
+	}
+
+	for {
+		select {
+		case event, ok := <-watcher.Events:
+			if !ok {
+				return
+			}
+
+			if isIgnoredFileEvent(event.Name, event.Op) {
+				utils.Debug("Ignoring file event: %s (%s)", event.Name, event.Op.String())
+				continue
+			}
+
+			// Handle file change events that should trigger a restart
+			isSignificantEventForRestart := event.Has(fsnotify.Write) || event.Has(fsnotify.Create) || event.Has(fsnotify.Remove) || event.Has(fsnotify.Rename)
+
+			if isSignificantEventForRestart {
+				utils.Info(2, "File change detected: %s (%s)", event.Name, event.Op.String())
+
+				// Re-add import dependencies to catch any new dependencies
+				if err := addImportDependenciesToWatcher(watcher, watchedDirs, fsys); err != nil {
+					utils.Warning("could not update import dependencies: %v", err)
+				}
+
+				if restartTimer != nil {
+					restartTimer.Stop()
+				}
+				restartTimer = time.AfterFunc(debounceDuration, func() {
+					restartChan <- struct{}{}
+				})
+			}
+
+		case err, ok := <-watcher.Errors:
+			if !ok {
+				return
+			}
+			utils.Warning("File watcher error: %v", err)
+		case <-ctx.Done():
+			if restartTimer != nil {
+				restartTimer.Stop()
+			}
+			return
+		}
+	}
+}
+
+// FileWatcherSetup interface allows for dependency injection of file watching functionality
+type FileWatcherSetup interface {
+	SetupFileWatcher(fsys afero.Fs) (*fsnotify.Watcher, string, error)
+}
+
+type RealFileWatcherSetup struct{}
```

Comment:
I would move some of the args of `runFileWatcher` to its own internal state. For eg.

```suggestion
type RealFileWatcher struct {
	watcher     *fsnotify.Watcher
	watchedDirs map[string]bool
	RestartChan chan struct{}
}
```


Then `SetupFileWatcher` should also be renamed to an idiomatic constructor.

```go
func NewFileWatcher(watchedPath string, fsys afero.Fs) (FileWatcher, error)
```